### ISSUE TITLE
Caching

### ISSUE THREAD

**Issue by snapshotpl** [zendframework/zend-expressive-fastroute#9]
FastRoute has cache feature https://github.com/nikic/FastRoute#caching . It's possible to intergate it into expressive?

**Comment by weierophinney** [comment]
Potentially, yes; we already support injecting a callback for generating the dispatcher instance, which allows alternate implementations. The issue is that both generateUri and match currently call injectRoutes, which you wouldn't want to do when facing is in place. 

As such, this might need to be done as an alternate implementation.

**Comment by pine3ree** [comment]
@snapshotpl, @weierophinney 
Thanks, this is a matter i wanted to discuss as well.
see <a href="https://github.com/nikic/FastRoute/issues/47" target="_blank">nikic/FastRoute/issues/47</a>
In web SAPI environments caching the generated regex is the way to go.
A possible (1st thought) solution would be modifying FastRouterRouter::addRoute(Route $route), avoiding adding to routesToInject if a caching-dispatcher is used and cached routes are succesfully loaded.
kind regards

**Comment by weierophinney** [comment]
@pine3ree —

> A possible (1st thought) solution would be modifying FastRouterRouter::addRoute(Route $route), avoiding adding to routesToInject if a caching-dispatcher is used and cached routes are succesfully loaded.

The question is really the second part of that statement: does the caching dispatcher have methods you can call to determine if routes are present? If it does, then this should be fairly trivial; if not, then it's a nightmare.

Additionally, we likely want to have a way to enable/disable caching based on environment. In development, where you may be adding, updating, or removing routes, you'll not want caching, but in production you will.

**Comment by pine3ree** [comment]
@weierophinney 
Actually there is no CachedDispatcher class in FastRoute, so we would need to implement all the functionality here.
Dispatching cached routes in FastRoute is achieved via a simple function (but i think the author provided it just as an example) located in <a href="https://github.com/nikic/FastRoute/blob/master/src/functions.php" target="_blank">FastRoute functions.php file</a>. It simply returns the desired dispatcher (which one is configurable via the $options parameter) feeded with data loaded from cache (when enabled and when generated data is loaded and valid).

i believe that in this bridge we would need to do something like:
- add a property to enable/disable caching (set via configuration)
- add a property for cached data
- if caching is enabled try to load data from cached file
- if cached data is not null skip adding to `routesToInject` (thus skipping calling `RouteCollector::addRoute(..)`) and feed the dispatcher
- if caching is enable but cached data is => persist data to file at the end of `injectRoutes()` 

so, maybe not a nightmare but also not a quick solution. The problem is that caching means to skip adding routes via the route collector which implictly calls the dataGenerator and this has to be done in FastRouteRouter, it cannot be moved to an external factory class.

kind regards

**Comment by weierophinney** [comment]
If either of you want to work up a pull request, I'll happily review it. :smile:

**Comment by pine3ree** [comment]
@weierophinney 
I hope i will have some spare time during next week to look into it :smile:

**Comment by pine3ree** [comment]
@weierophinney 

Hello Matthew, 
I added the caching feature <a href="https://github.com/pine3ree/zend-expressive-fastroute/blob/caching/src/FastRouteRouter.php" target="_blank">https://github.com/pine3ree/zend-expressive-fastroute/blob/patch-1/src/FastRouteRouter.php</a>

It's working fine locally, but i still need to add unit tests...

##### a few notes:
- I'll keep it separate (as it is a different feature) from PR7. In case it's ok, i believe it's better to complete the discussion and eventually the merging of PR7 before this (and i will have to rebase)
- in the FastRouterRouter factory we can pass options from $config...my naming choices to save a bit of FastRoute verbiage:
  - `$config['fastroute']['cache_enabled']` => `$options['cacheEnabled']`
  - `$config['fastroute']['cache_file']` => `$options['cacheFile']`
- cache filename: i used `.php.cache` as file extensions. I've been using json encoded data in cache files lately, having `.json.cache`, `.php.cache` helps me remember which solution i chose.

#### How it works:
1. pass $options to the constructor. If caching is enabled (try to) load dispatchData from cache
2. when injecting routes always build the route map (mandatory fro generateUri), but skip feeding the route collector if valid data was loaded from cache
3. match: the dispatcher will use cache data either from cache or freshly generated by the data generator via the route collector.

##### ideas/incompatibilities
- should we lazy load cached data inside injectRoutes or in an init() method called before injectRoutes?
- when using cached routes, any route added programmatically after the first request of dispatchData will not be either used nor cached, so it can be considered as an aut-aut choice: all the routes needs to added via configuration when using caching.

**Comment by snapshotpl** [comment]
@pine3ree great! Can you create PR in this project?

**Comment by pine3ree** [comment]
@snapshotpl 
I am going to wait for @weierophinney 's review and probably for PR #7 to be merged, so that i can issue a new clean pr without the need to rebase. 
I also have a small local commit which adds a private initialize method for loading more clearly the dispatch data when cache is enabled, as explained in one previous comment:

``` php
   //...
    /**
     * Control variable to make sure cached data is loaded only once
     *
     * @var bool
     */
    private $initialized = false;
    //...
    /**
     * Initialize the instance.
     *
     * @return void
     */
    private function initialize()
    {
        if (true === $this->initialized) {
            return;
        }

        if ($this->cacheEnabled) {
            $this->loadDispatchData();
        }

        $this->initialized = true;
    }
```

i call this method before injectRoutes() in match and generateUri

**Comment by pine3ree** [comment]
@snapshotpl  I added a PR #16  for review by mwof.
You are welcome to test it. (You need to create the router instance with the option `cacheEnabled` set to true, and optionally a different `cacheFile` path, default is `data/cache/fastroute.php.cache`)
kind regards

edited: renamed `cacheEnabled`  and `cacheFile`  to `cache_enabled` and `cache_file`

**Comment by pine3ree** [comment]
@snapshotpl 

the [PR](https://github.com/zendframework/zend-expressive-fastroute/pull/16) for fast-route caching support has been merged into 1.3.0 develop. I believe we can close this feature request issue.

kind regards
